### PR TITLE
feat (jkube-kit/common) : Add utility class to decompress archive files (#2456)

### LINKED ISSUE #2456: Add utility class to decompress archive files
### Component

JKube Kit

### Task description

Part of #439 

Related to #2454

As a JKube developer, I want to be able to let JKube extract a downloaded tar-gzipped or zip archive containing a Buildpacks CLI (pack) for my platform and system architecture to the specified directory, so that I can invoke a consistent version of the CLI tool (see #2453).

The required archive types for decompression should be:
- [x] tar gzipped archives `.tgz`
- [x] zip archives `.zip`

### Implementation

Utility should be implemented in the `org.eclipse.jkube.kit.common.archive` package of the `jkube-kit-common` module.

The utility should provide a `extractArchive(File inputFile, File targetDirectory)`.

For the Buildpacks CLI case, we should consider if the archive wraps a directory and see how to take care of this in order to locate the binary (probably belongs to #2454).

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fixes #2456

Adds ArchiveDecompressor utility class.
Contains an `extractArchive` method that extracts any File or InputStream containing an archive or a compressed archive into the provided target directory.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X] I have performed a self-review of my code
 - [X] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2471 (_2023-11-22T15:38:34Z_)

:gear: [JKube E2E Tests (6959547849)](https://github.com/jkubeio/ci/actions/runs/6959547849) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2471)](https://github.com/jkubeio/ci/actions/runs/6959547849/job/18937018088)

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeArchiveDecompressor.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  public static void extractArchive(File inputArchiveFile, File targetDirectory) throws IOException {
+    if (targetDirectory.exists()) {
+      FileUtil.cleanDirectory(targetDirectory);
+    }
+    Files.createDirectory(targetDirectory.toPath());
+    if (inputArchiveFile.getName().endsWith(".tgz")) {
+      extractTarArchive(inputArchiveFile, targetDirectory.toPath());
+    } else if (inputArchiveFile.getName().endsWith(".zip")) {
+      extractZipArchive(inputArchiveFile, targetDirectory.toPath());
+    }
```
Comment (manusa) [inline jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeArchiveDecompressor.java:49]: I believe this is not the optimal way to detect the compression algorithm and choose the decompression method (also it relies on filenames being in lower-case).

CompressorStreamFactory.createCompressorInputStream(originalInput);

might be a more suitable option

**Review by manusa — CHANGES_REQUESTED** [review]

File: jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeArchiveDecompressor.java (hunk 1 above)
Comment (rohanKanojia) [reply jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeArchiveDecompressor.java:49]: When I'm trying to pass file input stream to this method I'm getting mark not supported error:
`java.lang.IllegalArgumentException: Mark is not supported.`

Comment (manusa) [reply jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/archive/JKubeArchiveDecompressor.java:49]: Probably requires the Archive factory. Please check documentation, but I'm sure there are ways to detect the algorithms in commons-compress without relying on the file extension.

**Review by manusa — APPROVED** [review]

**Pull request merged**